### PR TITLE
treesitter: update runtime

### PR DESCRIPTION
**Pull request by vigoux**
Since https://github.com/tree-sitter/tree-sitter/pull/615, predicates are not parsed the same.
"Old" way of writing predicates is still supported.

This will also allow us to use newly added wildcard and repetitions for queries.

### CODE REVIEW

**Comment by tjdevries** [comment]
@vigoux is it literally just copy and paste to here? Do we have a script or anything to "easy-update" this for the future?

**Comment by vigoux** [comment]
Yeah for now just copy paste, I might try to make an auto update script though.

**Comment by tjdevries** [comment]
I think that'd be good. I actually need this PR soon anyway due to the predicate stuff you mentioned. For now I've just cherry-picked this commit onto mine locally. I think it might be worth to make a script that will do it though. Doesn't have to be the world's most amazingest script, but it'd be good if it was easy to do and we could be sure we didn't leave any extra files around, etc.

**Comment by h-michael** [comment]
It would be nice if the script could do this.
Also, regarding this PR, I think that you should include in the commit message what commit of the tree-sitter is included, like the commit(3ce9b05653a71efe2eeaeca6b6132d1227c367b7) of bfredl.

**Comment by vigoux** [comment]
Okay, the script is added, it is clearly not the best we have, be it works and does the job, you can call it like so `$0 <git commit sha you want to update to>` and it should work right away. Some manual patching is still needed to fix warnings though.

File: scripts/update-ts-runtime.sh
Comment (tjdevries) [inline scripts/update-ts-runtime.sh:23]: Do you need to `cd` here?

File: scripts/update-ts-runtime.sh
Comment (tjdevries) [inline scripts/update-ts-runtime.sh:37]: Can you put this before the tests? Tests should run last IMO

File: scripts/update-ts-runtime.sh
Comment (vigoux) [reply scripts/update-ts-runtime.sh:23]: That's something left from an earlier version, not used anymore, you are right.

File: scripts/update-ts-runtime.sh
Comment (tjdevries) [inline scripts/update-ts-runtime.sh:6]: Might want to `set -e` here

File: scripts/update-ts-runtime.sh
Comment (tjdevries) [inline scripts/update-ts-runtime.sh:14]: Echo here is unclear why it's being echoed. Tell me why it's echoing.

File: src/tree_sitter/alloc.h
Comment (tjdevries) [inline src/tree_sitter/alloc.h:25]: Is this section no longer needed? It is confusing to me why it was there but no longer is.

Comment (vigoux) [reply src/tree_sitter/alloc.h:25]: Actually, I keeped it last time because I thought it was mandatory, but visibly it is not anymore...
I can add it if you want though but I'm not sure of its usefullness...

**Comment by tjdevries** [comment]
What changes did you have to do manually to get it to work after running the script? (Wanted to post here so other people can see it later).

**Comment by tjdevries** [comment]
Also, another thing I was thinking was to make sure that any docs / tests / examples have the predicate syntax updated as well. I didn't check to see if there was anything that needed that or not.

File: src/tree_sitter/alloc.h
Comment (tjdevries) [reply src/tree_sitter/alloc.h:25]: Are any of those `#DEFINE`s used anywhere? Just like grepping through the source or anything like that.

Comment (tjdevries) [reply src/tree_sitter/alloc.h:25]: cc @bfredl if you're around.

**Comment by vigoux** [comment]
It works out of the box, but to fix warnings, here is the patch : 
```diff
diff --git a/src/tree_sitter/query.c b/src/tree_sitter/query.c
index 1ffc5ad6e..199960668 100644
--- a/src/tree_sitter/query.c
+++ b/src/tree_sitter/query.c
@@ -259,7 +259,7 @@ static void stream_scan_identifier(Stream *stream) {
  * CaptureListPool
  ******************/

-static CaptureListPool capture_list_pool_new() {
+static CaptureListPool capture_list_pool_new(void) {
   return (CaptureListPool) {
     .empty_list = array_new(),
     .usage_map = UINT32_MAX,
@@ -315,7 +315,7 @@ static void capture_list_pool_release(CaptureListPool *self, uint16_t id) {
  * SymbolTable
  **************/

-static SymbolTable symbol_table_new() {
+static SymbolTable symbol_table_new(void) {
   return (SymbolTable) {
     .characters = array_new(),
     .slices = array_new(),
@@ -752,7 +752,7 @@ static TSQueryError ts_query__parse_pattern(
       array_push(&branch_step_indices, start_index);
       array_push(&self->steps, query_step__new(0, depth, false));
     }
-    array_pop(&self->steps);
+    (void)array_pop(&self->steps);
 
     // For all of the branches except for the last one, add the subsequent branch as an
     // alternative, and link the end of the branch to the current end of the steps.
```

**Comment by vigoux** [comment]
> Also, another thing I was thinking was to make sure that any docs / tests / examples have the predicate syntax updated as well. I didn't check to see if there was anything that needed that or not.

So it is where things get interesting because both predicates syntaxes are still valid for now, so do we migrate to the newer syntax or do we keep support for both ?

**Comment by tjdevries** [comment]
I think just move to new one. If we get a lot of requests for old syntax, we can try and add it in.

**Comment by tjdevries** [comment]
@vigoux Do you have any tree sitter stuff that you can test on this build to make sure that it works beyond just the tests? I built this on my machine and seems to be working for most basic stuff.

**Comment by vigoux** [comment]
I did some tests with `nvim-treesitter` and everything seems to work fine.

**Comment by vigoux** [comment]
The patch isn't that trivial anymore though, here is the updated (and reverted) one :
```diff
diff --git a/src/tree_sitter/alloc.h b/src/tree_sitter/alloc.h
index 2229995bd..c8fe6c6e6 100644
--- a/src/tree_sitter/alloc.h
+++ b/src/tree_sitter/alloc.h
@@ -9,20 +9,7 @@ extern "C" {
 #include <stdbool.h>
 #include <stdio.h>
 
-#include "nvim/memory.h"
-
-#if 1
-
-static inline bool ts_toggle_allocation_recording(bool value) {
-  return false;
-}
-
-#define ts_malloc xmalloc
-#define ts_calloc xcalloc
-#define ts_realloc xrealloc
-#define ts_free xfree
-
-#elif defined(TREE_SITTER_TEST)
+#if defined(TREE_SITTER_TEST)
 
 void *ts_record_malloc(size_t);
 void *ts_record_calloc(size_t, size_t);
diff --git a/src/tree_sitter/query.c b/src/tree_sitter/query.c
index 1ffc5ad6e..199960668 100644
--- a/src/tree_sitter/query.c
+++ b/src/tree_sitter/query.c
@@ -259,7 +259,7 @@ static void stream_scan_identifier(Stream *stream) {
  * CaptureListPool
  ******************/
 
-static CaptureListPool capture_list_pool_new(void) {
+static CaptureListPool capture_list_pool_new() {
   return (CaptureListPool) {
     .empty_list = array_new(),
     .usage_map = UINT32_MAX,
@@ -315,7 +315,7 @@ static void capture_list_pool_release(CaptureListPool *self, uint16_t id) {
  * SymbolTable
  **************/
 
-static SymbolTable symbol_table_new(void) {
+static SymbolTable symbol_table_new() {
   return (SymbolTable) {
     .characters = array_new(),
     .slices = array_new(),
@@ -752,7 +752,7 @@ static TSQueryError ts_query__parse_pattern(
       array_push(&branch_step_indices, start_index);
       array_push(&self->steps, query_step__new(0, depth, false));
     }
-    (void)array_pop(&self->steps);
+    array_pop(&self->steps);
 
     // For all of the branches except for the last one, add the subsequent branch as an
     // alternative, and link the end of the branch to the current end of the steps.
diff --git a/src/tree_sitter/unicode.h b/src/tree_sitter/unicode.h
index 2ab51c2a3..0fba56a61 100644
--- a/src/tree_sitter/unicode.h
+++ b/src/tree_sitter/unicode.h
@@ -10,8 +10,8 @@ extern "C" {
 
 #define U_EXPORT
 #define U_EXPORT2
-#include "./unicode/utf8.h"
-#include "./unicode/utf16.h"
+#include "unicode/utf8.h"
+#include "unicode/utf16.h"
 
 static const int32_t TS_DECODE_ERROR = U_SENTINEL;
 
diff --git a/src/tree_sitter/unicode/umachine.h b/src/tree_sitter/unicode/umachine.h
index bbf6ef9c8..9195824d5 100644
--- a/src/tree_sitter/unicode/umachine.h
+++ b/src/tree_sitter/unicode/umachine.h
@@ -43,7 +43,7 @@
 /* which are contained in the platform-specific file platform.h             */
 /*==========================================================================*/
 
-#include "./ptypes.h" /* platform.h is included in ptypes.h */
+#include "unicode/ptypes.h" /* platform.h is included in ptypes.h */
 
 /*
  * ANSI C headers:
@@ -443,6 +443,6 @@ typedef int32_t UChar32;
  */
 #define U_SENTINEL (-1)
 
-#include "./urename.h"
+#include "unicode/urename.h"
 
 #endif
diff --git a/src/tree_sitter/unicode/utf16.h b/src/tree_sitter/unicode/utf16.h
index b54792244..9fd7d5c8a 100644
--- a/src/tree_sitter/unicode/utf16.h
+++ b/src/tree_sitter/unicode/utf16.h
@@ -34,9 +34,9 @@
 #ifndef __UTF16_H__
 #define __UTF16_H__
 
-#include "./umachine.h"
+#include "unicode/umachine.h"
 #ifndef __UTF_H__
-#   include "./utf.h"
+#   include "unicode/utf.h"
 #endif
 
 /* single-code point definitions -------------------------------------------- */
diff --git a/src/tree_sitter/unicode/utf8.h b/src/tree_sitter/unicode/utf8.h
index 3b37873e3..bb0013037 100644
--- a/src/tree_sitter/unicode/utf8.h
+++ b/src/tree_sitter/unicode/utf8.h
@@ -34,9 +34,9 @@
 #ifndef __UTF8_H__
 #define __UTF8_H__
 
-#include "./umachine.h"
+#include "unicode/umachine.h"
 #ifndef __UTF_H__
-#   include "./utf.h"
+#   include "unicode/utf.h"
 #endif
 
 /* internal definitions ----------------------------------------------------- */
```

**Comment by tjdevries** [comment]
I wonder if any of those small changes can be upstreamed. Obviously not the `ifdef` for nvim specific memory things, but the other small ones (like using void).

I'm not a C wizard, so I don't know if that introduces anything bad / unexpected upstream.

**Comment by tjdevries** [comment]
Maybe make an issue on tree-sitter and see what the thought is there (feel free to tag me)

**Comment by vigoux** [comment]
Maybe I'll open a PR directly, so that I get a review and they see if the changes are correct or not.

**Comment by tjdevries** [comment]
That sounds good. Can you tag me in PR so I can follow it?

**Comment by vigoux** [comment]
I will update once again once the upstream PR is merged, so that we don't have to fix those warnings on our side, and only patch for strictly neovim related things.

**Comment by vigoux** [comment]
It should be all good now, @tjdevries you can check this out if you want

**Review by tjdevries — APPROVED** [review]
@teto or @jamessan Do either of you have an opinion on this PR? It LGTM and I'm happy to merge, just wanted to see if anyone else had an opinion before I do.

**Comment by jamessan** [comment]
> I think it might be worth to make a script that will do it though. Doesn't have to be the world's most amazingest script, but it'd be good if it was easy to do and we could be sure we didn't leave any extra files around, etc.

Out of curiousity, is there any reason we didn't use git-subtree to pull in the tree-sitter source?

**Comment by teto** [comment]
can't find a reference in https://github.com/neovim/neovim/pull/10124 .

**Comment by tjdevries** [comment]
We don't pull it in exactly as they have it (there are some modifications we have in the source to build for us) and the structure of the code is different as well (we move some of their includes and other items to different locations in our build source). I think that's why.

**Comment by jamessan** [comment]
> can't find a reference to submodules in #10124 .

There was a reference to submodules, which Justin (rightly, IMHO) rejected due to their complexity.  However, git-subtree is a different beast and probably would have fit well.

> there are some modifications we have in the source to build for us

Ideally we could upstream those, or something that enables our use cases without requiring patches to tree-sitter.  This is going to make it more difficult for distributions to package the newer Neovim, since bundling is generally discouraged (or forbidden).

> the structure of the code is different as well (we move some of their includes and other items to different locations in our build source)

That seems like something we should be able to handle on our side of the build, or just not done if we had shoved tree-sitter in `third-party/` like everything else.

**Comment by jamessan** [comment]
Regardless, none of that should affect this PR.  The changes here look fine, albeit some more forethought initially would have avoided some of the changes in this PR.

**Comment by tjdevries** [comment]
We did upstream several of the small changes that we had to make in the past. The main one is that we patch in different memory functions it appears (we patch in using `nvim/memory` and `#define` some overrides to the functions they use. I'm not sure how that would be upstreamed.

Perhaps, as you mention, we should discuss this in a different issue though :laughing:

**Comment by teto** [comment]
the bundling issue mentioned by @jamessan is probably something we need to fix before 0.5. I had never heard of `git subtree` before but will try to play with it to get a better understanding. Anyway let's unblock our treesitter @vigoux for now.

**Comment by vigoux** [comment]
I will for sure look into it later, and try to see what I can do. For now though, my priority will be adding new features like language injection.

**Pull request merged**
